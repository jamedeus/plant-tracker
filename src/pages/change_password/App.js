import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [oldPassword, setOldPassword] = useState('');
    const [newPassword1, setNewPassword1] = useState('');
    const [newPassword2, setNewPassword2] = useState('');
    const [oldPasswordIncorrect, setOldPasswordIncorrect] = useState(false);
    const [newPasswordError, setNewPasswordError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post old and new password to backend
        const response = await fetch("/accounts/change_password/submit/", {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                "X-CSRFToken": Cookies.get('csrftoken')
            },
            credentials: "include",
            body: new URLSearchParams({
                old_password: oldPassword,
                new_password1: newPassword1,
                new_password2: newPassword2,
            }),
        });
        // Redirect to overview if logged in successfully
        if (response.ok) {
            window.location.href = '/';
        // Show correct error message if login failed
        } else {
            const errors = await response.json();
            if (errors.errors.old_password) {
                setOldPasswordIncorrect(true);
            }
            if (errors.errors.new_password2) {
                setNewPasswordError(errors.errors.new_password2[0]);
            }
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Change Password"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Old password</span>
                    </div>
                    <input
                        name="old_password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            oldPasswordIncorrect && "input-error"
                        )}
                        value={oldPassword}
                        onInput={() => setOldPasswordIncorrect(false)}
                        onChange={(e) => setOldPassword(e.target.value)}
                    />
                </label>
                {oldPasswordIncorrect && (
                    <span className="text-error">
                        Old password incorrect
                    </span>
                )}
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">New password</span>
                    </div>
                    <input
                        name="new_password1"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword1}
                        onChange={(e) => setNewPassword1(e.target.value)}
                    />
                </label>
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Confirm new password</span>
                    </div>
                    <input
                        name="new_password2"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword2}
                        onInput={() => setNewPasswordError(false)}
                        onChange={(e) => setNewPassword2(e.target.value)}
                    />
                </label>
                {newPasswordError && (
                    <span className="text-error">
                        {newPasswordError}
                    </span>
                )}

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                >
                    Change password
                </button>
            </form>
        </div>
    );
}

export default App;
