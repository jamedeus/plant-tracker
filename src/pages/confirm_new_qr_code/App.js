import React, { useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import { sendPostRequest, parseDomContext } from 'src/util';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';
import PlantDetails from 'src/components/PlantDetails.js';
import GroupDetails from 'src/components/GroupDetails.js';
import { openErrorModal } from 'src/components/ErrorModal';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faXmark, faCheck } from '@fortawesome/free-solid-svg-icons';

const Layout = ({ type, thumbnailUrl, handleConfirm }) => {
    return (
        <div className="flex flex-col gap-8 text-center my-auto">
            <p className="text-lg font-bold">
                Is this the new QR code for your {type}?
            </p>
            {thumbnailUrl && (
                <div className="mx-auto p-4 bg-base-200 rounded-3xl">
                    <img
                        className="max-h-[50vh] rounded-xl object-contain"
                        src={thumbnailUrl}
                    />
                </div>
            )}

            {/* Confirm/cancel buttons */}
            <div className="flex gap-4 mx-auto mb-8">
                <button
                    className="btn btn-error btn-square text-white"
                    onClick={() => window.location.href = "/"}
                    title="Don't change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faXmark} />
                </button>
                <button
                    className="btn btn-success btn-square text-white"
                    onClick={handleConfirm}
                    title="Change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faCheck} />
                </button>
            </div>
        </div>
    );
};

Layout.propTypes = {
    type: PropTypes.oneOf([
        'plant',
        'group'
    ]).isRequired,
    thumbnailUrl: PropTypes.string,
    handleConfirm: PropTypes.func.isRequired
};

function App() {
    // Load UUIDs and plant/group state from django template context
    const type = parseDomContext("type");
    const instance = parseDomContext("instance");
    const newUuid = parseDomContext("new_uuid");

    // Reload if user navigates to page by pressing back button (change QR code
    // request likely expired, replace with register or manage page)
    useEffect(() => {
        const handleBackButton = async (event) => {
            if (event.persisted) {
                window.location.reload();
            }
        };

        // Add listener on mount, remove on unmount
        window.addEventListener('pageshow', handleBackButton);
        return () => {
            window.removeEventListener('pageshow', handleBackButton);
        };
    }, []);

    const handleConfirm = async () => {
        const response = await sendPostRequest(
            '/change_uuid',
            {
                uuid: instance.uuid,
                new_id: newUuid
            }
        );
        // Reload page if changed successfully
        if (response.ok) {
            window.location.reload();
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    };

    // Plant/group details card shown when title is clicked
    const DetailsDropdown = useMemo(() => (
        <div className="details-card">
            <div className="card-body">
                {type === 'plant' ? (
                    <PlantDetails
                        species={instance.species}
                        pot_size={instance.pot_size}
                        description={instance.description}
                    />
                ) : (
                    <GroupDetails
                        location={instance.location}
                        description={instance.description}
                    />
                )}
            </div>
        </div>
    ), []);

    return (
        <div className="container flex flex-col mx-auto h-screen">
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title={instance.display_name}
                titleOptions={DetailsDropdown}
            />
            <Layout
                type={type}
                thumbnailUrl={instance.thumbnail}
                handleConfirm={handleConfirm}
            />
        </div>
    );
}

export default App;
