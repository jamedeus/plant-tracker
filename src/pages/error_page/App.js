import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';

function App({ errorMessage }) {
    return (
        <div
            className="container flex flex-col full-screen mx-auto items-center gap-16"
            data-testid="error-page"
        >
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title="Error"
                showScanButton={false}
            />

            <div className="flex flex-col text-center text-lg mt-[15vh]">
                {errorMessage}
            </div>

            <Link to="/" className='btn btn-accent' discover="none">
                Go to Overview
            </Link>
        </div>
    );
}

export default App;

App.propTypes = {
    errorMessage: PropTypes.string.isRequired
};
