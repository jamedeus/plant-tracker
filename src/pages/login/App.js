import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const [showError, setShowError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post FormData to backend
        const formData = new FormData(formRef.current);
        const response = await fetch('/accounts/login/', {
            method: 'POST',
            body: formData,
            headers: {
                'Accept': 'application/json, text/plain, */*',
                "X-CSRFToken": Cookies.get('csrftoken')
            }
        });
        // Redirect if logged in successfully
        if (response.ok) {
            // Redirect to url in querystring if present
            const params = new URL(window.location.href).searchParams;
            if (params.get('next')) {
                window.location.href = params.get('next');
            // Redirect to overview if no querystring
            } else {
                window.location.href = '/';
            }
        // Show error text if login failed
        } else {
            setShowError(true);
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Login"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Username</span>
                    </div>
                    <input
                        name="username"
                        type="text"
                        autoCapitalize="off"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={username}
                        onInput={() => setShowError(false)}
                        onChange={e => setUsername(e.target.value)}
                    />
                </label>
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">Password</span>
                    </div>
                    <input
                        name="password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={password}
                        onInput={() => setShowError(false)}
                        onChange={e => setPassword(e.target.value)}
                    />
                </label>

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                    disabled={username.length < 1 || password.length < 1}
                >
                    Login
                </button>

                {showError && (
                    <span className="text-error text-center">
                        Invalid username or password
                    </span>
                )}
            </form>
        </div>
    );
}

export default App;
