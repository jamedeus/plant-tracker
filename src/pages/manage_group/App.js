import React, { useState, useRef, useEffect, useCallback, useMemo } from 'react';
import PropTypes from 'prop-types';
import { localToUTC } from 'src/timestampUtils';
import { sendPostRequest, parseDomContext, pastTense } from 'src/util';
import EditModal from 'src/components/EditModal';
import GroupDetailsForm from 'src/forms/GroupDetailsForm';
import Navbar from 'src/components/Navbar';
import DatetimeInput from 'src/components/DatetimeInput';
import { showToast } from 'src/components/Toast';
import { useTheme } from 'src/context/ThemeContext';
import DetailsCard from 'src/components/DetailsCard';
import GroupDetails from 'src/components/GroupDetails';
import PlantsCol from 'src/components/PlantsCol';
import AddPlantsModal, { openAddPlantsModal } from './AddPlantsModal';
import RemovePlantsModal, { openRemovePlantsModal } from './RemovePlantsModal';
import ChangeQrModal, { openChangeQrModal } from 'src/components/ChangeQrModal';
import { openErrorModal } from 'src/components/ErrorModal';

// Buttons used to add events to all selected plants
const PlantEventButtons = ({ editing, setEditing, handleWater, handleFertilize }) => {
    const addEventTimeInput = useRef(null);

    // Handler for water button
    const water = async () => {
        handleWater(localToUTC(addEventTimeInput.current.value));
    };

    // Handler for fertilize button
    const fertilize = async () => {
        handleFertilize(localToUTC(addEventTimeInput.current.value));
    };

    if (editing) {
        return (
            <>
                <div
                    className="flex mx-auto mb-4"
                    data-testid="addEventTimeInput"
                >
                    <DatetimeInput inputRef={addEventTimeInput} />
                </div>
                <div className="flex">
                    <button
                        className="btn btn-outline mx-auto"
                        onClick={() => setEditing(false)}
                    >
                        Cancel
                    </button>
                    <button
                        className="btn btn-outline btn-info mx-auto"
                        onClick={water}
                    >
                        Water
                    </button>
                    <button
                        className="btn btn-outline btn-success mx-auto"
                        onClick={fertilize}
                    >
                        Fertilize
                    </button>
                </div>
            </>
        );
    } else {
        return (
            <div className="flex">
                <button
                    className="btn btn-outline mx-auto"
                    onClick={() => setEditing(true)}>
                    Manage
                </button>
            </div>
        );
    }
};

PlantEventButtons.propTypes = {
    editing: PropTypes.bool.isRequired,
    setEditing: PropTypes.func.isRequired,
    handleWater: PropTypes.func.isRequired,
    handleFertilize: PropTypes.func.isRequired
};

// Top-left menu button contents
const MenuOptions = () => {
    // Get toggle theme option from context
    const { ToggleThemeOption } = useTheme();

    return (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <li><a onClick={openAddPlantsModal}>
                Add plants
            </a></li>
            <li><a onClick={openRemovePlantsModal}>
                Remove plants
            </a></li>
            <li><a onClick={openChangeQrModal}>
                Change QR code
            </a></li>
            <ToggleThemeOption />
        </>
    );
};

// Contents of dropdown shown when group name clicked in header
const DetailsDropdown = ({ location, description }) => {
    return (
        <DetailsCard>
            <GroupDetails
                location={location}
                description={description}
            />
        </DetailsCard>
    );
};

DetailsDropdown.propTypes = {
    location: PropTypes.string,
    description: PropTypes.string
};

function App() {
    // Load context set by django template
    const [group, setGroup] = useState(() => {
        return parseDomContext("group");
    });
    const [plantDetails, setPlantDetails] = useState(() => {
        return parseDomContext("details");
    });
    // Contains list of objects with name and uuid of every plant in database
    // Should NOT be mutated except by pageshow listener below
    const [options, setOptions] = useState(() => {
        return parseDomContext("options");
    });

    // Array of plant objects that are not archived or already in group
    const addPlantsModalOptions = useMemo(() => {
        const existing = plantDetails.map(plant => plant.uuid);
        return options.filter(
            plant => !existing.includes(plant.uuid) && !plant.archived
        );
    }, [plantDetails, options]);

    // Request new state from backend if user navigates to page by pressing
    // back button (may be outdated if user clicked plant and made changes)
    useEffect(() => {
        const handleBackButton = async (event) => {
            if (event.persisted) {
                const response = await fetch(`/get_group_state/${group.uuid}`);
                if (response.ok) {
                    const data = await response.json();
                    setGroup(data['group']);
                    setPlantDetails(data['details']);
                    setOptions(data['options']);
                } else {
                    // Reload page if failed to get new state (group deleted)
                    window.location.reload();
                }
            }
        };

        window.addEventListener('pageshow', handleBackButton);

        return () => {
            window.removeEventListener('pageshow', handleBackButton);
        };
    }, []);

    // Create state to track whether selecting plants from list
    const [selectingPlants, setSelectingPlants] = useState(false);

    // FormRef for FilterColumn used to add events to subset of plants in group
    const selectedPlantsRef = useRef(null);

    // Returns array of selected plant UUIDs parsed from FilterColumn form
    const getSelectedPlants = () => {
        const selected = new FormData(selectedPlantsRef.current);
        return Array.from(selected.keys());
    };

    // Ref to access timestamp input used by water all/fertilize all
    const addEventAllTimeInput = useRef(null);

    // Create ref to access edit details form
    const editDetailsRef = useRef(null);

    const submitEditModal = useCallback(async () => {
        const payload = Object.fromEntries(
            new FormData(editDetailsRef.current)
        );
        payload["group_id"] = group.uuid;
        console.log(payload);

        const response = await sendPostRequest('/edit_group', payload);
        if (response.ok) {
            // Update plant state with new values from response
            const data = await response.json();
            setGroup(prevGroup => {
                return { ...prevGroup, ...data };
            });
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    }, []);

    // Takes array of plant UUIDs, removes archived plants and returns
    const removeArchivedPlants = (selected) => {
        return selected.filter(uuid => {
            const plant = plantDetails.find(plant => plant.uuid === uuid);
            if (plant && !plant.archived) {
                return uuid;
            }
        });
    };

    // Handler for "Water All" and "Fertilize All" buttons
    const addEventAll = async (eventType) => {
        const timestamp = localToUTC(addEventAllTimeInput.current.value);
        // Post eventType, UUIDs of all plants that aren't archived, and
        // timestamp to backend endpoint
        await bulkAddPlantEvents(
            eventType,
            removeArchivedPlants(plantDetails.map(plant => plant.uuid)),
            timestamp
        );
    };

    // Handler for water button under plant cards
    const waterSelected = async (timestamp) => {
        // Prevent watering archived plants
        const selected = removeArchivedPlants(getSelectedPlants());
        await bulkAddPlantEvents('water', selected, timestamp);
        setSelectingPlants(false);
    };

    // Handler for fertilize button under plant cards
    const fertilizeSelected = async (timestamp) => {
        // Prevent fertilizing archived plants
        const selected = removeArchivedPlants(getSelectedPlants());
        await bulkAddPlantEvents('fertilize', selected, timestamp);
        setSelectingPlants(false);
    };

    // Creates event with specified type and timestamp for every plant in
    // selectedIds (array of UUIDs)
    const bulkAddPlantEvents = async (eventType, selectedIds, timestamp) => {
        const payload = {
            plants: selectedIds,
            event_type: eventType,
            timestamp: timestamp
        };
        const response = await sendPostRequest('/bulk_add_plant_events', payload);
        if (response.ok) {
            showToast(`All plants ${pastTense(eventType)}!`, 'blue', 5000);
            const data = await response.json();
            updatePlantTimestamps(data.plants, timestamp, eventType);
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    };

    // Map eventType taken by bulk_add_plant_events to the plantDetails state
    // key that should be updated when an event is successfully created
    const eventTypeMap = {
        water: "last_watered",
        fertilize: "last_fertilized"
    };

    // Called by bulkAddPlantEvents to update plant water/fertilize timestamps
    const updatePlantTimestamps = (updatedPlants, timestamp, eventType) => {
        let newPlantDetails = [];
        const lastEvent = eventTypeMap[eventType];
        plantDetails.forEach(plant => {
            // Update lastEvent timestamp if UUID in JSON response
            if (updatedPlants.includes(plant.uuid)) {
                plant[lastEvent] = getMostRecent(plant[lastEvent], timestamp);
            }
            newPlantDetails.push(plant);
        });
        setPlantDetails(newPlantDetails);
    };

    // Takes 2 ISO 8601 timestamps, returns most recent
    const getMostRecent = (oldTime, newTime) => {
        // Return new if old is null (ie plant had no water events before)
        if (!oldTime) {
            return newTime;
        } else if (newTime > oldTime) {
            return newTime;
        } else {
            return oldTime;
        }
    };

    // Handler for add button in AddPlantsModal, takes array of UUIDs
    const addPlants = useCallback(async (selected) => {
        const payload = {
            group_id: group.uuid,
            plants: selected
        };
        const response = await sendPostRequest(
            '/bulk_add_plants_to_group',
            payload
        );
        if (response.ok) {
            // Add objects in response to plantDetails state
            const data = await response.json();
            setPlantDetails([...plantDetails, ...data.added]);
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    }, [plantDetails]);

    // Handler for remove button in RemovePlantsModal, takes array of UUIDs
    const removePlants = useCallback(async (selected) => {
        const payload = {
            group_id: group.uuid,
            plants: selected
        };
        const response = await sendPostRequest(
            '/bulk_remove_plants_from_group',
            payload
        );
        if (response.ok) {
            // Remove UUIDs in response from plantDetails
            const data = await response.json();
            setPlantDetails(plantDetails.filter(
                plant => !data.removed.includes(plant.uuid)
            ));
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    }, [plantDetails]);

    // Top left corner dropdown options
    const DropdownMenuOptions = useMemo(() => {
        return <MenuOptions />;
    }, []);

    // Group details card shown when title is clicked
    const GroupDetailsDropdown = useMemo(() => {
        return (
            <DetailsDropdown
                location={group.location}
                description={group.description}
            />
        );
    }, [group])

    return (
        <div className="container flex flex-col mx-auto mb-8">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title={group.display_name}
                titleOptions={GroupDetailsDropdown}
            />

            <DatetimeInput inputRef={addEventAllTimeInput} />
            <div className="flex mx-auto mb-8">
                <button
                    className="btn btn-info m-2"
                    onClick={() => addEventAll('water')}
                >
                    Water All
                </button>
                <button
                    className="btn btn-success m-2"
                    onClick={() => addEventAll('fertilize')}
                >
                    Fertilize All
                </button>
            </div>

            <PlantsCol
                plants={plantDetails}
                editing={selectingPlants}
                formRef={selectedPlantsRef}
                storageKey={`group-${group.uuid}`}
            >
                <PlantEventButtons
                    editing={selectingPlants}
                    setEditing={setSelectingPlants}
                    handleWater={waterSelected}
                    handleFertilize={fertilizeSelected}
                />
            </PlantsCol>

            <EditModal title="Edit Details" onSubmit={submitEditModal}>
                {useMemo(() => (
                    <GroupDetailsForm
                        formRef={editDetailsRef}
                        name={group.name}
                        location={group.location}
                        description={group.description}
                    />
                ), [group])}
            </EditModal>

            <AddPlantsModal
                options={addPlantsModalOptions}
                addPlants={addPlants}
            />

            <RemovePlantsModal
                plantDetails={plantDetails}
                removePlants={removePlants}
            />

            <ChangeQrModal
                uuid={group.uuid}
            />
        </div>
    );
}

export default App;
