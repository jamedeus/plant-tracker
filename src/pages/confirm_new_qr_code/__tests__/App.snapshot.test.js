import createMockContext from 'src/testUtils/createMockContext';
import App from '../App';
import { mockContext } from './mockContext';

describe('App', () => {
    it('matches snapshot when changing plant QR code', () => {
        // Create mock state objects
        createMockContext('type', 'plant');
        createMockContext('instance', mockContext.plant);
        createMockContext('new_uuid', mockContext.new_uuid);
        createMockContext('user_accounts_enabled', true);

        // Render App, confirm matches snapshot
        const component = render(<App />);
        expect(component).toMatchSnapshot();
    });

    it('matches snapshot when changing group QR code', () => {
        // Create mock state objects
        createMockContext('type', 'group');
        createMockContext('instance', mockContext.group);
        createMockContext('new_uuid', mockContext.new_uuid);
        createMockContext('user_accounts_enabled', true);

        // Render App, confirm matches snapshot
        const component = render(<App />);
        expect(component).toMatchSnapshot();
    });
});
