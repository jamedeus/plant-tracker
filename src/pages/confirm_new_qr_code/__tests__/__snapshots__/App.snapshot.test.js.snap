// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot when changing group QR code 1`] = `
<div>
  <div
    class="container flex flex-col mx-auto h-screen"
  >
    <div
      class="navbar bg-base-100 mb-4 sticky top-0 z-99"
    >
      <div
        class="dropdown justify-start min-w-12"
      >
        <div
          aria-label="Navigation menu"
          class="btn btn-ghost btn-circle size-12"
          role="button"
          tabindex="0"
        >
          <svg
            class="size-5"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h7"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <ul
          class="dropdown-content menu z-90 shadow-sm bg-base-300 rounded-box mt-3"
          tabindex="0"
        >
          <li>
            <a
              href="/"
            >
              Overview
            </a>
          </li>
          <li>
            <a
              href="/accounts/profile/"
            >
              User profile
            </a>
          </li>
          <li>
            <button>
              Light mode
            </button>
          </li>
        </ul>
      </div>
      <div
        class="mx-auto shrink min-w-0"
      >
        <div
          class="dropdown dropdown-center w-full"
        >
          <a
            class="btn btn-ghost w-full px-0 text-nowrap"
            role="button"
            tabindex="0"
          >
            <span
              class="px-4 leading-[2.75rem] truncate"
              style="font-size: 14px;"
            >
              Test Group
            </span>
          </a>
          <div
            class="dropdown-content z-1 flex"
            tabindex="0"
          >
            <div
              class="details-card"
            >
              <div
                class="card-body"
              >
                <p
                  class="flex"
                >
                  <span
                    class="font-semibold break-keep mr-4"
                  >
                    Location:
                  </span>
                  <span
                    class="ml-auto text-right line-clamp-1 break-all"
                  >
                    Nowhere
                  </span>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="justify-end min-w-12"
      />
    </div>
    <div
      class="flex flex-col gap-8 text-center my-auto"
    >
      <p
        class="text-lg font-bold"
      >
        Is this the new QR code for your 
        group
        ?
      </p>
      <div
        class="flex gap-4 mx-auto mb-8"
      >
        <a
          class="btn h-12 btn-error btn-square text-white"
          href="/"
          role="button"
          title="Don't change QR code"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-xmark size-6!"
            data-icon="xmark"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
              fill="currentColor"
            />
          </svg>
        </a>
        <button
          class="btn h-12 btn-success btn-square text-white"
          title="Change QR code"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check size-6!"
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M438.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L160 338.7 393.4 105.4c12.5-12.5 32.8-12.5 45.3 0z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`App matches snapshot when changing plant QR code 1`] = `
<div>
  <div
    class="container flex flex-col mx-auto h-screen"
  >
    <div
      class="navbar bg-base-100 mb-4 sticky top-0 z-99"
    >
      <div
        class="dropdown justify-start min-w-12"
      >
        <div
          aria-label="Navigation menu"
          class="btn btn-ghost btn-circle size-12"
          role="button"
          tabindex="0"
        >
          <svg
            class="size-5"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h7"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <ul
          class="dropdown-content menu z-90 shadow-sm bg-base-300 rounded-box mt-3"
          tabindex="0"
        >
          <li>
            <a
              href="/"
            >
              Overview
            </a>
          </li>
          <li>
            <a
              href="/accounts/profile/"
            >
              User profile
            </a>
          </li>
          <li>
            <button>
              Light mode
            </button>
          </li>
        </ul>
      </div>
      <div
        class="mx-auto shrink min-w-0"
      >
        <div
          class="dropdown dropdown-center w-full"
        >
          <a
            class="btn btn-ghost w-full px-0 text-nowrap"
            role="button"
            tabindex="0"
          >
            <span
              class="px-4 leading-[2.75rem] truncate"
              style="font-size: 14px;"
            >
              Test Plant
            </span>
          </a>
          <div
            class="dropdown-content z-1 flex"
            tabindex="0"
          >
            <div
              class="details-card"
            >
              <div
                class="card-body"
              >
                <p
                  class="flex"
                >
                  <span
                    class="font-semibold break-keep mr-4"
                  >
                    Species:
                  </span>
                  <span
                    class="ml-auto text-right line-clamp-1 break-all"
                  >
                    Calathea
                  </span>
                </p>
                <p
                  class="flex"
                >
                  <span
                    class="font-semibold break-keep mr-4"
                  >
                    Pot size:
                  </span>
                  <span
                    class="ml-auto text-right line-clamp-1 break-all"
                  >
                    4
                  </span>
                </p>
                <div
                  class="text-center"
                >
                  <p
                    class="font-semibold mt-3"
                  >
                    Description:
                  </p>
                  <p
                    class="text-sm line-clamp-6 break-anywhere"
                  >
                    This is a plant with a long description
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="justify-end min-w-12"
      />
    </div>
    <div
      class="flex flex-col gap-8 text-center my-auto"
    >
      <p
        class="text-lg font-bold"
      >
        Is this the new QR code for your 
        plant
        ?
      </p>
      <div
        class="mx-auto p-4 bg-base-200 rounded-3xl"
      >
        <img
          alt="Test Plant photo"
          class="max-h-[50vh] rounded-xl object-contain"
          draggable="false"
          src="/media/thumbnails/photo_thumb.jpg"
        />
      </div>
      <div
        class="flex gap-4 mx-auto mb-8"
      >
        <a
          class="btn h-12 btn-error btn-square text-white"
          href="/"
          role="button"
          title="Don't change QR code"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-xmark size-6!"
            data-icon="xmark"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
              fill="currentColor"
            />
          </svg>
        </a>
        <button
          class="btn h-12 btn-success btn-square text-white"
          title="Change QR code"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check size-6!"
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M438.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L160 338.7 393.4 105.4c12.5-12.5 32.8-12.5 45.3 0z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
